pkg/glusterblock/nodeserver.go

```diff
@@ -0,0 +1,172 @@
+package glusterblock
+
+import (
+	"context"
+	"os"
+	"path"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/gluster/gluster-csi-driver/pkg/utils"
+	"github.com/golang/glog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume/util"
+)
+
+// NodeServer struct of GlusterFS Block CSI driver with supported methods of CSI node server spec.
+type NodeServer struct {
+	*GfDriver
+}
+
+var mounter = mount.New("")
+
+// NodeStageVolume mounts the volume to a staging path on the node.
+func (ns *NodeServer) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+// NodeUnstageVolume unstages the volume from the staging path
+func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstageVolumeRequest) (*csi.NodeUnstageVolumeResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+// NodePublishVolume mounts the volume mounted to the staging path to the target path
+func (ns *NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) { //nolint: gocyclo
+	glog.V(2).Infof("received node publish volume request %+v", req)
+
+	err := ns.validateNodePublishVolumeReq(req)
+	if err != nil {
+		return nil, err
+	}
+
+	gs := req.GetVolumeContext()["glusterserver"]
+	volume := req.GetVolumeContext()["glustervol"]
+	hostPath, found := ns.Config.Mounts[volume]
+	if !found {
+		source := gs + ":" + volume
+		hostPath = path.Join(ns.Config.MntPathPrefix, volume)
+		err = utils.MountVolume(source, hostPath, "glusterfs", nil)
+		if err != nil {
+			return nil, err
+		}
+		ns.Config.Mounts[volume] = hostPath
+	}
+	srcPath := path.Join(hostPath, req.GetVolumeId())
+
+	if _, err = os.Stat(srcPath); os.IsNotExist(err) {
+		glog.Errorf("Block volume doesn't exist: %v", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	mo := req.GetVolumeCapability().GetMount().GetMountFlags()
+	if req.GetReadonly() {
+		mo = append(mo, ",ro")
+	}
+	targetPath := req.GetTargetPath()
+
+	// TODO fs should be argument
+	err = utils.MountVolume(srcPath, targetPath, "xfs", mo)
+	if err != nil {
+		return nil, err
+	}
+
+	return &csi.NodePublishVolumeResponse{}, nil
+}
+
+// NodeGetVolumeStats returns volume capacity statistics available for the
+// volume
+func (ns *NodeServer) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
+
+	// TODO need to implement volume status call
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+func (ns *NodeServer) validateNodePublishVolumeReq(req *csi.NodePublishVolumeRequest) error {
```

Comment:
this can be moved to common place

Reply:
I will do code refactor in separate patch